target_chains/cosmwasm/tools/src/pipeline.ts

```diff
@@ -0,0 +1,216 @@
+function readLineAsync(msg: string) {
+  const readline = createInterface({
+    input: process.stdin,
+    output: process.stdout,
+  });
+
+  return new Promise((resolve) => {
+    readline.question(msg, (userRes) => {
+      resolve(userRes);
+    });
+  });
+}
+
+type StateStore<T = any> = {
+  [stepId: string]: T;
+};
+// StateManager helps in getting and setting the state locally
+// It manipulates data in-memory and once the consumer has finished manipulating it
+// They need to commit the data to permanent storage using the commit method
+class StateManager<T> {
+  private readonly fileExt = ".json";
+  private readonly dirPath = "./tmp";
+
+  private readonly store: StateStore<T>;
+  private readonly filePath: string;
+
+  constructor(
+    private readonly stageId: string,
+    private readonly version: string
+  ) {
+    // dir check
+    if (!existsSync(this.dirPath)) {
+      mkdirSync(this.dirPath);
+    }
+
+    this.filePath = `${this.dirPath}/${this.stageId}-${this.version}${this.fileExt}`;
+    if (existsSync(this.filePath) === false) {
+      this.store = {};
+      return;
+    }
+    this.store = JSON.parse(readFileSync(this.filePath).toString());
+  }
+
+  // It gets the latest state for the given step
+  // the state after the last operation
+  getStepState(stepId: string): T | undefined {
+    return this.store[stepId];
+  }
+
+  // It sets the latest state for the given step
+  setSetState(stepId: string, state: T) {
+    this.store[stepId] = state;
+  }
+
+  // After all the in memory operations one can commit to the local file
+  // for permanent storage
+  commit() {
+    writeFileSync(this.filePath, JSON.stringify(this.store, null, 4));
+  }
+}
+
+// A step is defined as a method
+// it takes in a step id as a parameter to execute that particular step
+// Optionally it can take in a method `getResultOfPastStage` as a parameter
+// if it wants to access the result of the previous stages
+export type Step =
+  | ((
+      stepId: string,
+      // get the result of a past stage using it's id
+      // It will return the result for the same step id
+      // It will return undefined if the previous stage data has not been stored locally
+      // or if a future stage data is being asked
+      getResultOfPastStage: <Y>(stageId: string) => Y
+    ) => Promise<any>)
+  | ((stepId: string) => Promise<any>);
+
+// A step can fail. If the error is not handled it will crash the pipeline
+// We would like to store the result or reason locally too
+// For that purpose the response from a step is being wrapped in this
+type StepResult<T = any> =
+  | {
+      status: "rejected";
+      stepId: string;
+      reason: any;
+    }
+  | {
+      status: "fulfilled";
+      stepId: string;
+      result: T;
+    };
+
+// A stage will contain an identifier and a step method
+type Stage = {
+  stageId: string;
+  step: Step;
+};
+
+// A pipeline is conists of multiple stages
+// A stage contains of multiple steps
```

Comment:
I think having stages having multiple "equal" steps makes things very complex. I think it is because in first part you want to deploy every contract, then instantiate and so on?

I think having a pipeline executing a sequence of steps is easier and enough for us.